Moved ThriftClientHandler out of TabletServer for #1581

andrewglowacki opened the pull request:
Moved ThriftClientHandler into it's own file out of TabletServer. Also moved AssignmentHandler and UnloadTabletHandler into separate files (from TabletServer) since they are created by ThriftClientHandler and can no longer be inner classes.

ThriftClientHandler is still tightly coupled to TabletServer and can't really exist on it's own.

ctubbsii commented on server/tserver/src/main/java/org/apache/accumulo/tserver/ThriftClientHandler.java:
```diff
@@ -0,0 +1,1812 @@
+    super(server.getContext(), server.getWatcher(), server.getFileSystem());
+    this.server = server;
+    this.watcher = server.getWatcher();
+    this.fs = server.getFileSystem();
```
You don't actually need to get the "file system" from the tserver. You can get it from the ServerContext:

```suggestion
    this.fs = server.getContext().getVolumeManager();
```

ctubbsii commented on server/tserver/src/main/java/org/apache/accumulo/tserver/ThriftClientHandler.java:
```diff
@@ -0,0 +1,1812 @@
+    this.server = server;
+    this.watcher = server.getWatcher();
+    this.fs = server.getFileSystem();
+    this.security = server.getSecurity();
```
I think I saw that there was already a `.getSecurityOperation()` method.

ctubbsii commented on server/tserver/src/main/java/org/apache/accumulo/tserver/ThriftClientHandler.java:
```diff
@@ -0,0 +1,1812 @@
+  private final TabletServerLogger logger;
+
+  public ThriftClientHandler(TabletServer server) {
+    super(server.getContext(), server.getWatcher(), server.getFileSystem());
+    this.server = server;
+    this.watcher = server.getWatcher();
```
As stated earlier, you can get the stuff from the context and the super class:

```suggestion
    super(server.getContext(), new TransactionWatcher(server.getContext()), server.getContext().getVolumeManager());
    this.server = server;
    this.watcher = transactionWatcher;
```

ctubbsii commented on server/tserver/src/main/java/org/apache/accumulo/tserver/ThriftClientHandler.java:
```diff
@@ -0,0 +1,1812 @@
+
+    synchronized (unopenedTablets) {
+      synchronized (openingTablets) {
+        SortedMap<KeyExtent,Tablet> onlineTablets = server.getOnlineTablets();
+        synchronized (onlineTablets) {
```
This is changing the synchronization. Previously, it synchronized on the snapshot factory. Now, it's synchronizing on the snapshot. This kind of error is why I think it'd probably be better to use package-private references to TabletServer's own fields, rather than use getters and assign to new variables.

I saw you avoided this behavior change elsewhere by synchronizing on what is returned from your new method, `getOnlineTabletsRaw()`, but I think it's too easy to mistake these methods for one another.

Another reason to use the same reference is because it makes it easier to analyze the concurrency model by finding references the object in an IDE. If the same object is assigned to multiple references, trying to understand the concurrency model is more difficult, because you have to follow call stacks.

andrewglowacki replied:
Whoops, yea, the syncrhonziation isn't actually necessary since it's operating on the snapshot. Yea, as I mentioned in slack, the use of syncrhonzied colletions makes this a bit confusing and would be more straight forward to refactor them as concurrent collections later.

ctubbsii replied:
Perhaps... making them concurrent instead of having the explicit synchronization... might make it harder to understand the overall concurrency model, and the relationship between these objects. It might make more sense to put the objects into a single container that gives a consistent snapshot across all of the data structures at once. I'd have to think about it some more to have an informed opinion.

andrewglowacki commented on server/tserver/src/main/java/org/apache/accumulo/tserver/AssignmentHandler.java:
```diff
@@ -0,0 +1,243 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AssignmentHandler implements Runnable {
```
Makes sense, I wasn't sure whether or not to go public or package private here.

andrewglowacki commented on server/tserver/src/main/java/org/apache/accumulo/tserver/AssignmentHandler.java:
```diff
@@ -0,0 +1,243 @@
+
+  public AssignmentHandler(TabletServer server, KeyExtent extent, int retryAttempt) {
+    this.server = server;
+    this.onlineTablets = server.getOnlineTabletsRaw();
+    this.unopenedTablets = server.getUnopenedTablets();
+    this.openingTablets = server.getOpeningTablets();
```
I like this proposition a better since it'll make it a lot easier to ensure the refactor is a 1-to-1 mapping. Does it make sense for these methods to exist down the road though to ease testing of ThriftClientHandler by itself? Or do you think TabletServer and ThriftClientHandler shouldn't need to be tested in isolation?

ctubbsii replied:
I don't think they will need to be tested in isolation for now. They may make sense to be added in future, but for now, I think keeping the code smaller will help enable subsequent improvements.

andrewglowacki commented on server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java:
```diff
@@ -307,6 +194,10 @@ public TabletServerMinCMetrics getMinCMetrics() {
     return mincMetrics;
   }
 
+  public TabletServerUpdateMetrics getUpdateMetrics() {
+    return updateMetrics;
+  }
+
```
Makes sense.

andrewglowacki commented on server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java:
```diff
@@ -495,2084 +386,218 @@ private static long jitter() {
-      for (TColumn tcolumn : columns) {
-        columnSet.add(new Column(tcolumn));
-      }
+  protected TransactionWatcher getWatcher() {
+    return watcher;
+  }
```
Oh, yes, thanks! I caught this while I was refactoring 1.9 earlier but I missed it this time around.

ctubbsii replied:
There's actually another room for improvement (not in this PR... but subsequently): the ClientServiceHandler (parent class of ThriftClientHandler) only needs one parameter, not three: the server context. But, that's for a subsequent change.

andrewglowacki commented on server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java:
```diff
@@ -495,2084 +386,218 @@ private static long jitter() {
-          readaheadThreshold, executionHints);
-      scanSession.scanner =
-          tablet.createScanner(new Range(range), scanParams, scanSession.interruptFlag);
+  protected String getLockID() {
+    return lockID;
+  }
```
Makes sense. Unrelated, but I was wondering why this and some other variables weren't volatile, should they be?

ctubbsii replied:
Not sure at this time. I'd have to investigate how they are used.

keith-turner replied:
> Unrelated, but I was wondering why this and some other variables weren't volatile, should they be?

If its not final and accessed by multiple threads outside of any lock then it seems like it should be.  That could be a follow on PR.

keith-turner replied:
> However, I think this one should stay, because it provides access to a non-final field

There is at least one other field that is non final and that is clientAddress.

andrewglowacki commented on server/tserver/src/main/java/org/apache/accumulo/tserver/ThriftClientHandler.java:
```diff
@@ -0,0 +1,1812 @@
+public class ThriftClientHandler extends ClientServiceHandler implements TabletClientService.Iface {
+
+  private static final Logger log = LoggerFactory.getLogger(ThriftClientHandler.class);
+  private static final long MAX_TIME_TO_WAIT_FOR_SCAN_RESULT_MILLIS = 1000;
+  private static final long RECENTLY_SPLIT_MILLIES = 60 * 1000;
+  private final TabletServer server;
+  private final WriteTracker writeTracker = new WriteTracker();
+  private final RowLocks rowLocks = new RowLocks();
```
Thanks! :-)

andrewglowacki commented:
Thanks for the detailed review and great comments! 

Unfortunately, while I was making changes, I rebased master onto my branch (to get my vscode changes), and it re-wrote about 4-5 commits from master by mistake. I backed those out, but I had to force push since I had pushed them to my remote branch.

ctubbsii commented:
Merging master onto your branch would have been okay if you need to pick up changes from there. It's just the force-push that breaks things... it causes all review comments to become automatically obsolete, and it makes it extremely difficult, if not impossible, to view what changed from the last code review. So, a code reviewer basically has to start over, reviewing from scratch.

andrewglowacki commented:
Good to know next time, sorry again! I didn't change the original commit, so all of the comments should still be applicable.

ctubbsii commented:
> Good to know next time, sorry again! I didn't change the original commit, so all of the comments should still be applicable.

No worries.

To be honest, I used to prefer force pushing a lot... and didn't understand why it mattered that much when others, such as @keith-turner, expressed a preference to avoid it. Then, I started doing a lot more code reviews, and then it became a pain point. So now I've become one of the ones advising people not to force push. Tables turned :smiley_cat:

keith-turner commented:
@andrewglowacki would you be able to highlight the places where these changes may be doing something different or new?

andrewglowacki commented:
@keith-turner sorry, what changes do you mean? The force push didn't add/remove anything from the original commit, it just removed the commits that I accidently pulled in when I rebased master. This is what happened:

1. Original commit: 8d99cdc
2. Reviewed by @ctubbsii 
3. I rebased master which pulled in and modified the next 4 commits
4. Pushed this to my remote branch before I realized the 4 commits were modified by the rebase
5. Reset the head to 8d99cdc
6. Force push the branch to remove the 4 commits
7. Committed and push my changes from the review in 1bf3ba5

keith-turner commented:
@andrewglowacki this was my first time to look at the PR this morning, so the force push did not impact me. I think a lot of the changes are only something like adding `server.` with no other changes to code.   I was just wondering if you could point me to more substantial changes to review, like a changing a foreach loop to a java stream.  One possible way to do this is to comment on the PR where these more substantial changes are.   I am just looking for a way to review this more quickly, but I don't want to take much more of your time.  So if this would not be quick for you, don't worry about it.

andrewglowacki commented:
@keith-turner I see, great! No problem, I don't mind pointing things out. I'll do anything to make it easier to review, especially since it's such a big change.

Unfortunately there's quite a few changes, however virtually all of them are either:
- prepending variables that were owned by TabletServer with 'server.' such as: 
`sessionManager.unreserveSession(scanSession);` 
to:
`server.sessionManager.unreserveSession(scanSession);`

- prepending methods that were owned by TabletServer with 'server.' such as:
`long totalQueued = updateTotalQueuedMutationSize(additionalMutationSize);`
to:
`long totalQueued = server.updateTotalQueuedMutationSize(additionalMutationSize);`

- changing private variables to package-private in TabletServer.java such as:
`private final TabletServerResourceManager resourceManager;`
to:
`final TabletServerResourceManager resourceManager;`

In addition to making these types of changes to ThriftClientHandler, they were also necessary in AssignmentHandler and UnloadTabletHandler for the same reasons.

keith-turner commented:
@andrewglowacki the three changes you just mentioned are great and I approve of those.  I was just looking for pointers to anything else you may know of.  I plan to take a look over the changes this afternoon.

ctubbsii commented:
@andrewglowacki I will try to take a look at this Tuesday, and merge it in if I don't see any issues with it. I gave myself a break for the weekend. :smiley_cat:

andrewglowacki commented:
@ctubbsii no problem, thanks!

keith-turner commented on server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java:
```diff
@@ -2278,301 +529,13 @@ public void enqueueMasterMessage(MasterMessage m) {
-  }
-
-  private void acquireRecoveryMemory(KeyExtent extent) {
+  protected void acquireRecoveryMemory(KeyExtent extent) {
```
```suggestion
 void acquireRecoveryMemory(KeyExtent extent) {
```

keith-turner commented on server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java:
```diff
@@ -2278,301 +529,13 @@ public void enqueueMasterMessage(MasterMessage m) {
   }
 
-  private void releaseRecoveryMemory(KeyExtent extent) {
+  protected void releaseRecoveryMemory(KeyExtent extent) {
```
```suggestion
 void releaseRecoveryMemory(KeyExtent extent) {
```

keith-turner commented on server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java:
```diff
@@ -2228,7 +479,7 @@ private void splitTablet(Tablet tablet) {
   }
 
-  private TreeMap<KeyExtent,TabletData> splitTablet(Tablet tablet, byte[] splitPoint)
+  protected TreeMap<KeyExtent,TabletData> splitTablet(Tablet tablet, byte[] splitPoint)
```
```suggestion
 TreeMap<KeyExtent,TabletData> splitTablet(Tablet tablet, byte[] splitPoint)
```

keith-turner commented on server/tserver/src/main/java/org/apache/accumulo/tserver/TabletServer.java:
```diff
@@ -493,1649 +374,19 @@ private static long jitter() {
-      if (session == null) {
-        throw new NoSuchScanIDException();
-      }
+  protected String getLockID() {
```
```suggestion
 String getLockID() {
```

andrewglowacki commented:
@keith-turner thanks for the review! I will make these changes.

keith-turner commented:
> @keith-turner thanks for the review! I will make these changes.

No problem.  I did the refactoring locally in Eclipse and then compared that to your branch.  That made the diff much smaller and allowed me to see all of the changes you made that were not automatic.

andrewglowacki commented:
> I did the refactoring locally in Eclipse and then compared that to your branch.  That made the diff much smaller and allowed me to see all of the changes you made that were not automatic.

Good idea! Yes, it was pretty difficult to discern otherwise.

ctubbsii commented:
All tests passed (well, SuspendedTabletsIT.crashAndResumeTserver
 experienced an unrelated timeout, but passed when I re-ran it; it's probably transient, performance related on the build system). I'm merging this.

ctubbsii commented:
Thanks for the PR @andrewglowacki ; just to reiterate a previous comment: https://github.com/apache/accumulo/pull/1583#issuecomment-611674874

andrewglowacki commented:
No problem. Thanks for the reviews and help!
